File: reference_code/world_model/model/reference_line_model.py
import sys
import sympy as sym
from sympy import *

import numpy as np

from core.ocp_model import *

file_path = os.path.abspath(__file__)
dir_path = os.path.dirname(file_path)
solver_path = os.path.dirname(dir_path)
gen_dir = solver_path + "/model/reference_line_model"
if len(sys.argv) > 1:
    gen_dir = sys.argv[1]

ocp_model = OptimalControlProblem("ReferenceLineModel")

# define state symbols
x, y, theta, kappa = ocp_model.add("state", "x, y, theta, kappa")

# define control symbols
dkappa, slack_offset = ocp_model.add("ctrl", "dkappa, slack_offset")
# dkappa = ocp_model.add("ctrl", "dkappa")

# define system equation
ocp_model.subject(dot(x) == cos(theta))
ocp_model.subject(dot(y) == sin(theta))
ocp_model.subject(dot(theta) == kappa)
ocp_model.subject(dot(kappa) == dkappa)

# define cost parameters
xr, yr, thetar = ocp_model.add("param", "xr, yr, thetar")
(
    l_weight,
    kappa_weight,
    dkappa_weight,
    v_weight,
    slack_offset_weight,
    l_weight_terminal,
    kappa_weight_terminal
) = ocp_model.add("param",
                  "l_weight, kappa_weight, dkappa_weight, v_weight, slack_offset_weight, l_weight_terminal, kappa_weight_terminal"
                  )

# cost
l = -sin(thetar) * (x - xr) + cos(thetar) * (y - yr)
l_x = 0.5 * l_weight * l**2 + 0.5 * kappa_weight * kappa**2
l_u = 0.5 * dkappa_weight * dkappa**2 + 0.5 * slack_offset_weight * slack_offset**2

# l_u = 0.5 * dkappa_weight * dkappa**2


ocp_model.add("cost", l_x, l_u)

# define terminal cost parameters
l_t = 0.5 * l_weight_terminal * l**2 + 0.5 * kappa_weight_terminal * kappa**2

ocp_model.add("cost", l_t, type=StageType.TERMINAL)

# define constraints
#ll, lu = ocp_model.add("param", "ll", "lu")
sll, slu = ocp_model.add("param", "sll", "slu")
KappaLowerBound, KappaUpperBound = ocp_model.add("param",
                                                 "KappaLowerBound", "KappaUpperBound"
                                                 )
DKappaLowerBound, DKappaUpperBound = ocp_model.add("param",
                                                   "DKappaLowerBound", "DKappaUpperBound"
                                                   )

theta_offset = cos(theta - thetar)
#ocp_model.subject(ll <= l, l <= lu)
ocp_model.subject(0.5 <= theta_offset)
ocp_model.subject(sll <= l + slack_offset, l - slack_offset <= slu)
ocp_model.subject(kappa >= KappaLowerBound, kappa <= KappaUpperBound)
ocp_model.subject(dkappa <= DKappaUpperBound, dkappa >= DKappaLowerBound)

ocp_model.generate(gen_dir)

File: src/reference_code/world_model/model/reference_line_model/reference_line_model.cpp
#include "reference_line_model.h"

namespace gpal::pnc::planning {

ReferenceLineModel::ReferenceLineModel() : OptimalControlProblem("ReferenceLineModel", 4, 2, 0, 16, 4) {
  default_state_.emplace("x");
  default_state_.emplace("y");
  default_state_.emplace("theta");
  default_state_.emplace("kappa");

  default_ctrl_.emplace("dkappa");
  default_ctrl_.emplace("slack_offset");

  default_param_.emplace("xr", 0.0);
  default_param_.emplace("yr", 0.0);
  default_param_.emplace("thetar", 0.0);
  default_param_.emplace("l_weight", 0.0);
  default_param_.emplace("kappa_weight", 0.0);
  default_param_.emplace("dkappa_weight", 0.0);
  default_param_.emplace("v_weight", 0.0);
  default_param_.emplace("slack_offset_weight", 0.0);
  default_param_.emplace("l_weight_terminal", 0.0);
  default_param_.emplace("kappa_weight_terminal", 0.0);
  default_param_.emplace("sll", 0.0);
  default_param_.emplace("slu", 0.0);
  default_param_.emplace("KappaLowerBound", 0.0);
  default_param_.emplace("KappaUpperBound", 0.0);
  default_param_.emplace("DKappaLowerBound", 0.0);
  default_param_.emplace("DKappaUpperBound", 0.0);

  default_global_.emplace("cos_thetar");
  default_global_.emplace("sin_thetar");
  default_global_.emplace("theta_offset");
  default_global_.emplace("l");
  default_global_.setUpdater(std::make_shared<OcpVariable::Updater>([](const OcpStage& stage, OcpVariable* ptr_global) {
    // Determine stage variables
    const auto& x = stage.x(0);
    const auto& y = stage.x(1);
    const auto& theta = stage.x(2);
    const auto& xr = stage.params(0);
    const auto& yr = stage.params(1);
    const auto& thetar = stage.params(2);

    (*ptr_global)(0) = cos(thetar);
    (*ptr_global)(1) = sin(thetar);
    (*ptr_global)(2) = cos(theta - thetar);
    (*ptr_global)(3) = (*ptr_global)(0) * (y - yr) - (*ptr_global)(1) * (x - xr);
  }));
}

std::shared_ptr<Dynamics> ReferenceLineModel::createDynamics(OcpConfig::IntegratorType type) const {
  return std::make_shared<ReferenceLineModelDynamics>(type);
}

std::shared_ptr<CostFunction> ReferenceLineModel::createCostFunction() const {
  return std::make_shared<ReferenceLineModelCost<StageType::NORMINAL>>();
}

std::shared_ptr<CostFunction> ReferenceLineModel::createTerminalCostFunction() const {
  return std::make_shared<ReferenceLineModelCost<StageType::TERMINAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createConstraint() const {
  return std::make_shared<ReferenceLineModelConstraints<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createInitialConstraint() const {
  return std::make_shared<ReferenceLineModelConstraints<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createTerminalConstraint() const {
  return std::make_shared<ReferenceLineModelConstraints<StageType::TERMINAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createStateOnlyEqualities() const {
  return std::make_shared<ReferenceLineModelStateOnlyEqualities<StageType::NORMINAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createInitialStateOnlyEqualities() const {
  return std::make_shared<ReferenceLineModelStateOnlyEqualities<StageType::INITIAL>>();
}

std::shared_ptr<Constraints> ReferenceLineModel::createTerminalEqualities() const {
  return std::make_shared<ReferenceLineModelStateOnlyEqualities<StageType::TERMINAL>>();
}

std::shared_ptr<IPMHelper> ReferenceLineModel::createBaseIpmHelper() const {
  IPMHelper::Creater creater{
      {{StageType::NORMINAL, OcpConfig::FORWARD_EULER},
       []() {
         return std::make_shared<ReferenceLineModelIpmEvaluator<StageType::NORMINAL, OcpConfig::FORWARD_EULER>>();
       }},
      {{StageType::NORMINAL, OcpConfig::ERK4},
       []() { return std::make_shared<ReferenceLineModelIpmEvaluator<StageType::NORMINAL, OcpConfig::ERK4>>(); }},
      {{StageType::INITIAL, OcpConfig::FORWARD_EULER},
       []() {
         return std::make_shared<ReferenceLineModelIpmEvaluator<StageType::INITIAL, OcpConfig::FORWARD_EULER>>();
       }},
      {{StageType::INITIAL, OcpConfig::ERK4},
       []() { return std::make_shared<ReferenceLineModelIpmEvaluator<StageType::INITIAL, OcpConfig::ERK4>>(); }},
      {{StageType::TERMINAL, OcpConfig::FORWARD_EULER},
       []() { return std::make_shared<ReferenceLineModelIpmEvaluatorTerminal>(); }},
      {{StageType::TERMINAL, OcpConfig::ERK4},
       []() { return std::make_shared<ReferenceLineModelIpmEvaluatorTerminal>(); }}};
  return std::make_shared<IPMHelper>(std::move(creater));
}

REGIST_MODEL(ReferenceLineModel)
}  // namespace gpal::pnc::planning